zipline/utils/cme_exchange_calendar.py

```diff
@@ -0,0 +1,454 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from datetime import time
+from itertools import chain
+
+import numpy as np
+import pandas as pd
+from dateutil.relativedelta import (
+    MO,
+    TH,
+)
+from pandas import (
+    date_range,
+    DateOffset,
+    Timestamp,
+)
+from pandas.tseries.holiday import(
+    AbstractHolidayCalendar,
+    GoodFriday,
+    Holiday,
+    nearest_workday,
+    sunday_to_monday,
+    USLaborDay,
+    USPresidentsDay,
+    USThanksgivingDay,
+)
+from pandas.tseries.offsets import Day
+from pytz import timezone
+
+from zipline.utils.calendars import ExchangeCalendar
+
+# Useful resources for making changes to this file:
+# http://www.nyse.com/pdfs/closings.pdf
+# http://www.stevemorse.org/jcal/whendid.html
+
+MONDAY, TUESDAY, WEDNESDAY, THURSDAY, FRIDAY, SATURDAY, SUNDAY = range(7)
+
+US_CENTRAL = timezone('Americas/Chicago')
+CME_OPEN = time(17)
+CME_CLOSE = time(16)
+# CME_STANDARD_EARLY_CLOSE = time(13)
+# Does the market open or close on a different calendar day, compared to the
+# calendar day assigned by the exchang to this session?
+CME_OPEN_OFFSET = -1
+CME_CLOSE_OFFSET = 0
+
+# Closings
+USNewYearsDay = Holiday(
+    'New Years Day',
+    month=1,
+    day=1,
+    # When Jan 1 is a Sunday, NYSE observes the subsequent Monday.  When Jan 1
+    # Saturday (as in 2005 and 2011), no holiday is observed.
+    observance=sunday_to_monday
+)
+USMemorialDay = Holiday(
+    # NOTE: The definition for Memorial Day is incorrect as of pandas 0.16.0.
+    # See https://github.com/pydata/pandas/issues/9760.
+    'Memorial Day',
+    month=5,
+    day=25,
+    offset=DateOffset(weekday=MO(1)),
+)
+USMartinLutherKingJrAfter1998 = Holiday(
+    'Dr. Martin Luther King Jr. Day',
+    month=1,
+    day=1,
+    # The NYSE didn't observe MLK day as a holiday until 1998.
+    start_date=Timestamp('1998-01-01'),
+    offset=DateOffset(weekday=MO(3)),
+)
+USIndependenceDay = Holiday(
+    'July 4th',
+    month=7,
+    day=4,
+    observance=nearest_workday,
+)
+Christmas = Holiday(
+    'Christmas',
+    month=12,
+    day=25,
+    observance=nearest_workday,
+)
+
+# Half Days
+MonTuesThursBeforeIndependenceDay = Holiday(
+    # When July 4th is a Tuesday, Wednesday, or Friday, the previous day is a
+    # half day.
+    'Mondays, Tuesdays, and Thursdays Before Independence Day',
+    month=7,
+    day=3,
+    days_of_week=(MONDAY, TUESDAY, THURSDAY),
+    start_date=Timestamp("1995-01-01"),
+)
+FridayAfterIndependenceDayExcept2013 = Holiday(
```

Comment:
Most of these oddities were developed by hand by checking against the history of the NYSE.  Does the CME follow exactly the NYSE schedule?